### PR TITLE
Introduce unit tests

### PR DESCRIPTION
**Pull request by achumachenko-cloudinary**
Implemented modifications to allow unit testing and end-to-end testing of the script functionality.

### CODE REVIEW

File: cld-bulk.js [hunk 1]
```diff
@@ -0,0 +1,86 @@
+#!/usr/bin/env node
+/**
+ * @fileoverview This is the main entry point for the CLI tool.
+ * 
+ * 💡 Edit the `__input-to-api-payload.js` module to customize how CSV input is "translated" to Cloudinary API payload
+ * 
+ * Parsess command line parameters and invokes the main processing loop.
```
Comment (mgriff) [inline cld-bulk.js]: SP?  Parses

File: cld-bulk.js [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+
+//
+// Configure command line arguments shared across all commands
+//
+function yieldDefaultArgsCommand(program) {
+    const defaultArgsCommand = program.createCommand()
+        .requiredOption(
+            '-f, --from-csv-file <path>', 
+            'CSV file detailing assets to import',
+            cliHelpers.inpuFileMustExist)
```
Comment (mgriff) [inline cld-bulk.js]: incorrect method, should be `inputFileMustExist`

File: lib/input/cli-helpers.js [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+ */
+
+const fs = require('node:fs');
+const commander = require('commander');
+const logging = require('../output/logging');
+const reporting = require('../output/reporting');
+
+
+const MIN_CONCURRENT_OPERATIONS = 1;
+const MAX_CONCURRENT_OPERATIONS = 20;
```
Comment (mgriff) [inline lib/input/cli-helpers.js:12]: What made you pick 20? 

I like the idea of having a max but feel it could be a little higher, maybe 30.

I always use 20 for my migrations

File: lib/input/confirmation-routines.js [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+ *
+ * @async
+ * @function confirmOperationOptionsOrExit_Async
+ * @param {Object} migrationOptions - The set of migration parameters to be confirmed.
+ * @throws {UserDidNotConfirmError} - If the user does not confirm the operation parameters.
+ */
+async function confirmOperationOptionsOrExit_Async(migrationOptions) {
+    const migrationOptionsStr = util.inspect(migrationOptions, {depth: null, colors: true});
+    const migrationPrompt = 
+`❗️WARNING: This script will perform asset migration with the following parameters:
```
Comment (mgriff) [inline lib/input/confirmation-routines.js]: Should this be "... will perform the **bulk operation** with the ..." now that you've added support for update?

File: lib/main-loop.js [hunk 5]
```diff
@@ -0,0 +1,153 @@
...
+
+    // Using async generator to avoid loading the entire input file into memory
+    const inputRecordGeneratorAsync = csvReader.getRecordGenerator_Async(inputCsvFilePath);
+
+    // Using async.mapLimit to limit the number of concurrent operations
+    await async.mapLimit(inputRecordGeneratorAsync, maxConcurrentUploads, async (input) => {
+        let payload = null;
+        let response = null;
+        let summary = {
+            status: 'MIGRATED',
```
Comment (mgriff) [inline lib/main-loop.js:92]: Having "MIGRATED' as the default feels risky to me.

Should this not be set after we have the response?

I see there being 4 status':
- MIGRATED = New upload of an asset
- OVERWRITTEN = Public id already existed and `overwrite` was set to `true` (`overwritten` is in the response object)
- ALREADY EXISTED = Public id already existed and `overwrite` was set to `false` (`existing` is in the response object)
- FAILED = Error occurred

File: lib/output/progress.js [hunk 6]
```diff
@@ -61,7 +61,7 @@ async function init_Async(filePath) {
         fps: 5
     });
     _statusBar = _multiBar.create(totalCount, 0, init_stats, {
-        format: 'Attempted: {value} (✅{succeeded} /❌{failed}) out of {total}',
+        format: 'Attempted: {value} (✅{succeeded} /❌{failed}) out of estimated {total}',
```
Comment (mgriff) [inline lib/output/progress.js:64]: Just for my curiosity, why did you add an estimated here?

Is there a scenario where it doesn't get the actual total?

File: test/.resources/sample.jpg
Comment (mgriff) [inline test/.resources/sample.jpg:1]: Small thing but can we use the latest version of the logo 😄

**Review by mgriff — CHANGES_REQUESTED** [review]
I've added few suggestions/questions in line.

A more general comment on the folder structure, I personally think that it looks cleaner to have all the tests under one folder (rather that with the source files). I realise it can come down to personal choice but there are a few reasons I think we should separate them out:

- It separates out the core code from the tests files which makes it easier IMO for people to understand the roles of the individual files
-  It helps if you need to run a subset of tests, for example if the different test types are in subfolders of a root test folder, you can use `jest test/*/*` for all tests, and `jest test/unit/*` or `jest test/integration/*` for unit or full e2e tests.  
- This is the same approach that the Cloudinary JS SDK follows so would be more consistent across our code bases

File: cld-bulk.js (hunk 1 above)
Comment (achumachenko-cloudinary) [reply cld-bulk.js]: Fixed

File: cld-bulk.js (hunk 2 above)
Comment (achumachenko-cloudinary) [reply cld-bulk.js]: Great catch 🤘 ! Fixed.

File: lib/input/cli-helpers.js (hunk 3 above)
Comment (achumachenko-cloudinary) [reply lib/input/cli-helpers.js:12]: There isn't exact science there. Thought process is more along the lines of:

We recommend to start with 10: https://support.cloudinary.com/hc/en-us/articles/202520662-How-can-I-bulk-upload-my-images-
20 seems to be reasonable for most applications (and reduces chances of 420).
Those who know what they are doing and really need more can edit this definition  :) 

LMK what you think @mgriff

File: lib/input/confirmation-routines.js (hunk 4 above)
Comment (achumachenko-cloudinary) [reply lib/input/confirmation-routines.js]: Good eye! Changed

File: lib/main-loop.js (hunk 5 above)
Comment (achumachenko-cloudinary) [reply lib/main-loop.js:92]: This would need to be a more involved change. 
The tool had been used with this in place - think it is OK for now.

Created issue to address moving forward: https://github.com/cloudinary/cld-migrate-remote-assets-in-bulk/issues/9
LMK what you think @mgriff 

Cheers!

File: lib/output/progress.js (hunk 6 above)
Comment (achumachenko-cloudinary) [reply lib/output/progress.js:64]: For performance reasons (processing large CSV files) the scripts [when estimating number of records] does not parse CSV records but counts the number of lines in the file: https://github.com/cloudinary/cld-migrate-remote-assets-in-bulk/blob/6753cc29d0105d0c257cb6f83ae3a2bc1278e9dd/lib/output/progress.js#L52

The number of lines may be different (larger) from number of records in the file (for example, if some of the values are double-quote enclosed with newline characters in them.

File: test/.resources/sample.jpg
Comment (achumachenko-cloudinary) [reply test/.resources/sample.jpg:1]: Done. Updated logo to the current version.
"How you do anything is how you do everything" 🤙

**Comment by achumachenko-cloudinary** [comment]
RE:
> A more general comment on the folder structure, I personally think that it looks cleaner to have all the tests under one folder (rather that with the source files). I realise it can come down to personal choice but there are a few reasons I think we should separate them out:
> 
> * It separates out the core code from the tests files which makes it easier IMO for people to understand the roles of the individual files
> * It helps if you need to run a subset of tests, for example if the different test types are in subfolders of a root test folder, you can use `jest test/*/*` for all tests, and `jest test/unit/*` or `jest test/integration/*` for unit or full e2e tests.
> * This is the same approach that the Cloudinary JS SDK follows so would be more consistent across our code bases


I don't think it makes sense to mimic the setup of the SDK (we already use different testing framework jest vs mocha in SDK). 

My reasoning has been:
- This is a small app so it should be easy to make sense of where tests are
- You can still run specific tests `jest lib/input/*`

This doesn't seem to be essential aspect of the implementation so I prefer moving faster vs getting things exactly right.
LMK what you think @mgriff 

Appreciate all your time and the feedback! 🤘

Cheers!

File: lib/input/cli-helpers.js (hunk 3 above)
Comment (mgriff) [reply lib/input/cli-helpers.js:12]: Sounds good to me, we can always change it in the future if we get push back.

A scale up mechanism would also be a nice addition, one that handles 420s etc - one for the far future 😄

File: lib/main-loop.js (hunk 5 above)
Comment (mgriff) [reply lib/main-loop.js:92]: Agreed that it makes sense to move this to it's own issue.

File: lib/output/progress.js (hunk 6 above)
Comment (mgriff) [reply lib/output/progress.js:64]: Ah makes sense, thanks for the explanation 👍

**Review by mgriff — APPROVED** [review]
LGTM
